components/codec/number.rs

```diff
@@ -0,0 +1,2078 @@
+    #[bench]
+    fn bench_decode_u64_le_bytes_buf(b: &mut test::Bencher) {
+        use bytes::Buf;
+
+        let buf: Vec<u8> = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
+        b.iter(|| {
+            let mut cursor = ::std::io::Cursor::new(test::black_box(buf.as_slice()));
+            let v = cursor.get_u64_le();
+            test::black_box(v);
+            test::black_box(cursor.position());
+        });
+    }
+
+    /// Decode u64 little endian using `BufferNumberDecoder` over a `Cursor<&[u8]>`.
+    #[bench]
+    fn bench_decode_u64_le_buffer_decoder(b: &mut test::Bencher) {
+        use super::BufferNumberDecoder;
+
+        let buf: Vec<u8> = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
+        b.iter(|| {
+            let mut cursor = ::std::io::Cursor::new(test::black_box(buf.as_slice()));
+            let v = cursor.read_u64_le().unwrap();
+            test::black_box(v);
+            test::black_box(cursor.position());
+        });
+    }
+
+    /// Decode u64 little endian using the original implementation in TiKV.
+    #[bench]
+    fn bench_decode_u64_le_original(b: &mut test::Bencher) {
+        let buf: Vec<u8> = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
+        b.iter(|| {
+            let mut slice = test::black_box(buf.as_slice());
+            let v = original_decode_u64_le(&mut slice).unwrap();
+            test::black_box(v);
+            test::black_box(&slice);
+        });
+    }
+
+    const VARINT_SAMPLE: u64 = 0xDEADBEEFBAAD;
+    const VARINT_ENCODED: [u8; 7] = [173, 245, 190, 247, 219, 213, 55];
+
+    /// Encode u64 in VarInt using `NumberCodec`.
+    #[bench]
+    fn bench_encode_varint_number_codec(b: &mut test::Bencher) {
+        let mut buf: Vec<u8> = vec![0; super::MAX_VARINT64_LENGTH];
+        b.iter(|| {
+            let mut pos = 0;
+            let bytes = super::NumberCodec::encode_var_u64(
+                test::black_box(&mut buf[..]),
+                test::black_box(VARINT_SAMPLE),
+            );
+            pos += bytes;
+            test::black_box(&buf);
+            test::black_box(pos);
+        });
+    }
+
+    trait OldVarIntEncoder: byteorder::WriteBytesExt {
+        fn encode_var_u64(&mut self, mut v: u64) -> super::Result<()> {
+            while v >= 0x80 {
+                self.write_u8(v as u8 | 0x80)
+                    .map_err(|_| super::Error::BufferTooSmall)?;
+                v >>= 7;
+            }
+            self.write_u8(v as u8)
+                .map_err(|_| super::Error::BufferTooSmall)
+        }
+    }
+
+    impl<T: byteorder::WriteBytesExt> OldVarIntEncoder for T {}
+
+    /// Encode u64 in VarInt using original TiKV implementation.
+    #[bench]
+    fn bench_encode_varint_original(b: &mut test::Bencher) {
+        let mut buf: Vec<u8> = Vec::with_capacity(super::MAX_VARINT64_LENGTH);
+        b.iter(|| {
+            OldVarIntEncoder::encode_var_u64(
+                test::black_box(&mut buf),
+                test::black_box(VARINT_S
```

Comment:
Looks like 9 is too complicated for rust to understand. You can change it to nested loop like
```
let mut i = 0;
for _ in 0..3 {
    for _ in 0..3 {
        i += 1;
    }
}
```

Performance should not be worse than the folly's.

Reply:
This seems to be more tricky than a basic loop unrolling. There is no documentation that ensures that this unrolling will work as intended. In contrast, you can easily check current PR's correctness by looking at folly's implementation.

Reply:
It's too ugly to write all those duplicated codes by hand. When rust supports specify attributes to force unrolling a loop, it can be used, until then I think two for loop is the most clean and readable solution.

> check current PR's correctness

I thought this PR is going to improve performance. When changing the loop from `0..9` to nested loop, tikv's version performs better than folly's.